apisix/plugins/api-breaker.lua

```diff
@@ -0,0 +1,180 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local ngx = ngx
+local math = math
+local os = os
+local ipairs   = ipairs
+local core = require("apisix.core")
+local plugin_name = "api-breaker"
+
+local shared_buffer = ngx.shared['plugin-'.. plugin_name]
+
+local schema = {
+    type = "object",
+    properties = {
+        response_code = {
+            type = "integer",
+            minimum = 200,
+            maximum = 599,
+        },
+        unhealthy = {
+            type = "object",
+            http_statuses = {
+                type = "array",
+                minItems = 1,
+                items = {
+                    type = "integer",
+                    minimum = 500,
+                    maximum = 599,
+                },
+                uniqueItems = true,
+                default = {500}
```

Comment:
why default config not work? @membphis

Reply:
why you say it not work? I do not found any test case about this default value?

Reply:
test case as :
```
=== TEST 7: add plugin with default config value
--- config
    location /t {
        content_by_lua_block {
            local t = require("lib.test_admin").test
            local code, body = t('/apisix/admin/routes/1',
                ngx.HTTP_PUT,
                [[{
                    "plugins": {
                        "api-breaker": {
                            "response_code": 502,
                            "unhealthy": {
                                "failures": 3
                            },
                            "healthy": {
                                "successes": 3
                            }
                        }
                    },
                    "upstream": {
                        "nodes": {
                            "127.0.0.1:1988": 1
                        },
                        "type": "roundrobin"
                    },
                    "uri": "/test"
                }]]
                )
            if code >= 300 then
                ngx.status = code
            end
            ngx.say(body)
        }
    }
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]


=== TEST 8: default value
--- request
GET /test?r=500
--- error_code: 500
--- no_error_log
[error]
```

got err: 
```
t/plugin/api-breaker.t .. 53/? 
#   Failed test 'TEST 8: default value - status code ok'
#   at /usr/local/share/perl/5.26.1/Test/Nginx/Socket.pm line 935.
#          got: ''
#     expected: '500'

#   Failed test 'TEST 8: default value - pattern "[error]" should not match any line in error.log but matches line "2020/09/29 09:02:48 [error] 50729#50729: *2 failed to run header_filter_by_lua*: ./apisix/plugins/api-breaker.lua:84: bad argument #1 to 'ipairs' (table expected, got nil)" (req 0)
# stack traceback:
#       [C]: in function 'ipairs'
#       ./apisix/plugins/api-breaker.lua:84: in function 'is_unhealthy'
#       ./apisix/plugins/api-breaker.lua:171: in function 'phase_func'
#       ./apisix/init.lua:150: in function 'run_plugin'
#       ./apisix/init.lua:613: in function 'common_phase'
#       ./apisix/init.lua:621: in function 'http_header_filter_phase'
# '
#   at /usr/local/share/perl/5.26.1/Test/Nginx/Socket.pm line 1280.
# Looks like you failed 2 tests of 54.
t/plugin/api-breaker.t .. Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/54 subtests 
```

Reply:
let me confirm this bug. if I have any news, I'll reply here

Reply:
@liuhengloveyou your schema is wrong. you can use the common tools to check it first.

![image](https://user-images.githubusercontent.com/6814606/94648532-50405a00-0325-11eb-851d-e30a05ecba52.png)

Reply:
thand you a lot!